src/sentry/search/snuba/base.py

```diff
@@ -0,0 +1,732 @@
+                    # of groups moving around in the sort scoring underneath us,
+                    # so we at least want to protect against duplicates
+                    continue
+
+                group_score = group_to_score[group_id]
+                result_group_ids.add(group_id)
+                result_groups.append((group_id, group_score))
+
+        # TODO do we actually have to rebuild this SequencePaginator every time
+        # or can we just make it after we've broken out of the loop?
+        paginator_results = SequencePaginator(
+            [(score, id) for (id, score) in result_groups], reverse=True, **paginator_options
+        ).get_result(limit, cursor, known_hits=hits)
+
+        # break the query loop for one of three reasons:
+        # * we started with Postgres candidates and so only do one Snuba query max
+        # * the paginator is returning enough results to satisfy the query (>= the limit)
+        # * there are no more groups in Snuba to post-filter
+        if candidate_ids or len(paginator_results.results) >= limit or not more_results:
+            break
+
+    # HACK: We're using the SequencePaginator to mask the complexities of going
+    # back and forth between two databases. This causes a problem with pagination
+    # because we're 'lying' to the SequencePaginator (it thinks it has the entire
+    # result set in memory when it does not). For this reason we need to make some
+    # best guesses as to whether the `prev` and `next` cursors have more results.
+
+    if len(paginator_results.results) == limit and more_results:
+        # Because we are going back and forth between DBs there is a small
+        # chance that we will hand the SequencePaginator exactly `limit`
+        # items. In this case the paginator will assume there are no more
+        # results, so we need to override the `next` cursor's results.
+        paginator_results.next.has_results = True
+
+    if cursor is not None and (not cursor.is_prev or len(paginator_results.results) > 0):
+        # If the user passed a cursor, and it isn't already a 0 result `is_prev`
+        # cursor, then it's worth allowing them to go back a page to check for
+        # more results.
+        paginator_results.prev.has_results = True
+
+    metrics.timing("snuba.search.num_chunks", num_chunks)
+
+    groups = Group.objects.in_bulk(paginator_results.results)
+    paginator_results.results = [groups[k] for k in paginator_results.results if k in groups]
+
+    return paginator_results
+
+
+class SnubaSearchBackendBase(SearchBackend):
+    def query(
+        self,
+        projects,
+        environments=None,
+        sort_by="date",
+        limit=100,
+        cursor=None,
+        count_hits=False,
+        paginator_options=None,
+        search_filters=None,
+        date_from=None,
+        date_to=None,
+    ):
+        # print ("Running Original Backend Query...")
```

Comment:
Let's remove any leftover prints. If you want to keep them, switch them over to `logger.debug`